Replace negative results from LevelZero with NaN and avoid issues with high order bits in active time

cmcantalupo opened the pull request:
- For of the interfaces where the geopm::LevelZero class returns a double precision value, check for negative results and replace.
- Some of the geopm::LevelZero interfaces return integer values and these are unmodified.
- Fixes #3589

cmcantalupo committed:
Replace negative results from LevelZero with NaN

- The frequency API specifies that negative may be returned for invalid
- Fixes #3589

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

avilcheslopez commented on libgeopmd/src/LevelZero.cpp:
```diff
@@ -626,7 +633,7 @@ namespace geopm
                         GEOPM_ERROR_RUNTIME, "LevelZero::" + std::string(__func__) +
                         ": Sysman failed to get performance factor values", __LINE__);
-        return result;
+        return conert_nan(result);
```
Typo

bgeltz reviewed (commented):
Revisit the L0 docu around all our APIs and verify whether or not the negative check is needed.

bgeltz commented:
Also look at GPU_UTILIZATION and GPU_CORE_UTILIZATION re: the composite fields (i.e. GPU_ACTIVE_TIME and GPU_ACTIVE_TIME_TIMESTAMP).

The signal definitions for GPU_ACTIVE_TIME, GPU_ACTIVE_TIME_TIMESTAMP, and the CORE_ACTIVE flavors, all cast the underlying uint64_t to a double when a read happens.  If a rollover has occurred, this will result in bad math for the computed signal.

Related to #3550.

cmcantalupo committed:
Fix issues with LevelZero return values

- Do not throw for read_signal() calls for TIMESTAMP signals
- LevelZero does not give units for "TIME" values, do not scale.
- Ignore upper 32 bits of active time counter

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
Debug output from syslog:
``` 
Oct 18 12:57:52 NODENAME geopmd[PID]: DEBUG: derivative: 1.35277e+10 t = [3.81657e+09,4.05126e+09,4.05375e+09,4.05625e+09,4.05875e+09,4.06125e+09,4.06374e+09,4.06624e+09], y = [4.7416e+08,4.7416e+08,4.74473e+08,4.74473e+08,4.74473e+08,4.74768e+08,4.76796e+08,1.84467e+19]
Oct 18 12:57:52 NODENAME geopmd[PID]: DEBUG: derivative: 1.05553e+12 t = [4.05126e+09,4.05375e+09,4.05625e+09,4.05875e+09,4.06125e+09,4.06374e+09,4.06624e+09,4.06873e+09], y = [4.7416e+08,4.74473e+08,4.74473e+08,4.74473e+08,4.74768e+08,4.76796e+08,1.84467e+19,1.84467e+19]
Oct 18 12:57:52 NODENAME geopmd[PID]: DEBUG: derivative: 1.31942e+12 t = [4.05375e+09,4.05625e+09,4.05875e+09,4.06125e+09,4.06374e+09,4.06624e+09,4.06873e+09,4.07123e+09], y = [4.74473e+08,4.74473e+08,4.74473e+08,4.74768e+08,4.76796e+08,1.84467e+19,1.84467e+19,1.84467e+19]
Oct 18 12:57:52 NODENAME geopmd[PID]: DEBUG: derivative: 1.40735e+12 t = [4.05625e+09,4.05875e+09,4.06125e+09,4.06374e+09,4.06624e+09,4.06873e+09,4.07123e+09,4.07373e+09], y = [4.74473e+08,4.74473e+08,4.74768e+08,4.76796e+08,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19]
Oct 18 12:57:52 NODENAME geopmd[PID]: DEBUG: derivative: 1.31945e+12 t = [4.05875e+09,4.06125e+09,4.06374e+09,4.06624e+09,4.06873e+09,4.07123e+09,4.07373e+09,4.07622e+09], y = [4.74473e+08,4.74768e+08,4.76796e+08,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19]
Oct 18 12:57:53 NODENAME geopmd[PID]: DEBUG: derivative: 1.0556e+12 t = [4.06125e+09,4.06374e+09,4.06624e+09,4.06873e+09,4.07123e+09,4.07373e+09,4.07622e+09,4.07872e+09], y = [4.74768e+08,4.76796e+08,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19]
Oct 18 12:57:53 NODENAME geopmd[PID]: DEBUG: derivative: 6.15725e+11 t = [4.06374e+09,4.06624e+09,4.06873e+09,4.07123e+09,4.07373e+09,4.07622e+09,4.07872e+09,4.08122e+09], y = [4.76796e+08,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19,1.84467e+19]
```

Note:
```
>>> float(0xffffffffffffffff)
1.8446744073709552e+19
 ```

cmcantalupo commented on libgeopmd/src/LevelZeroDevicePool.cpp:
```diff
@@ -302,8 +318,10 @@ namespace geopm
+        uint64_t active_time = m_levelzero.active_time(dev_subdev_idx_pair.first, l0_domain,
+                                                       dev_subdev_idx_pair.second);
+
+        return convert_active_time(active_time, m_active_time_last[domain_idx], m_active_time_rollover[domain_idx]);
```
There needs to be another dimesion to m_active_time_lastand m_active_time_rollover: one for each l0_domain.

cmcantalupo replied:
https://github.com/geopm/geopm/pull/3615/commits/bce35cbab30dc36fc5d24d4fa8a0d58ea11efcee

cmcantalupo committed:
Fixup rollover cache so it is per LevelZero domain

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
```
$ cat session-big.cfg
TIME board 0
GPU_CORE_ACTIVITY gpu_chip 0
GPU_CORE_ACTIVITY gpu_chip 1
GPU_CORE_ACTIVITY gpu_chip 2
GPU_CORE_ACTIVITY gpu_chip 3
GPU_CORE_ACTIVITY gpu_chip 4
GPU_CORE_ACTIVITY gpu_chip 5
GPU_CORE_ACTIVITY gpu_chip 6
GPU_CORE_ACTIVITY gpu_chip 7
GPU_CORE_ACTIVITY gpu_chip 8
GPU_CORE_ACTIVITY gpu_chip 9
GPU_CORE_ACTIVITY gpu_chip 10
GPU_CORE_ACTIVITY gpu_chip 11
$ for ((i=0; i<10; i++)) do DATE=$(date +%FT%T%Z); ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 >& run_mpiexec-${DATE}.log & PID=$!; cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv & wait; cat $(ls -t *.csv | head -n1) | csvstat --null-value=nan | grep 'est value'; done
[1] 84608
[2] 84610
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         100.324
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        -0
	Largest value:         0.986
	Smallest value:        -0
	Largest value:         0.99
	Smallest value:        -0
	Largest value:         0.989
	Smallest value:        -0
	Largest value:         0.991
[1] 84977
[2] 84979
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.522
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.991
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         1.144
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
[1] 85316
[2] 85318
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.323
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.991
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.988
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
[1] 85628
[2] 85630
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.11
	Largest value:         98.717
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         1.121
	Smallest value:        0
	Largest value:         1.123
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.988
	Smallest value:        0
	Largest value:         0.991
[1] 86361
[2] 86363
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         99.123
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
[1] 86683
[2] 86685
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.723
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.992
[1] 87000
[2] 87002
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.724
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
[1] 87733
[2] 87735
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.112
	Largest value:         99.52
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.988
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.988
	Smallest value:        0
	Largest value:         0.991
[1] 88052
[2] 88054
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.522
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.993
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
[1] 88377
[2] 88379
[1]-  Done                    ./run_mpiexec.sh -n=1 -rpn=12 --ranks_per_socket=6 &> run_mpiexec-${DATE}.log
[2]+  Done                    cat session-big.cfg | geopmsession --print-header --pid ${PID} -p 0.2 -t 360000 > session-fix-${DATE}.csv
	Smallest value:        0.116
	Largest value:         98.123
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.99
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
	Smallest value:        0
	Largest value:         0.992
	Smallest value:        0
	Largest value:         0.987
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.986
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.989
	Smallest value:        0
	Largest value:         0.991
```

dannosliwcd commented on libgeopmd/src/LevelZeroIOGroup.cpp:
```diff
@@ -359,7 +359,7 @@ namespace geopm
                                                    geopm::LevelZero::M_DOMAIN_ALL);
                                   },
-                                  1 / 1e6
+                                  1
```
For this and the other time & timestamp signals:

Should we also update the description text to remove the "in seconds" notes since the L0 docs don't commit to specific units?

cmcantalupo replied:
https://github.com/geopm/geopm/pull/3615/commits/2a9b54d97ce1770d1c02448dd9dc770f9cba7902

cmcantalupo committed:
Update units definition for LevelZero activity

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on libgeopmd/src/LevelZeroDevicePool.cpp:
```diff
@@ -266,9 +266,23 @@ namespace geopm
-    uint64_t LevelZeroDevicePoolImp::active_time_timestamp(int domain,
-                                                           unsigned int domain_idx,
-                                                           int l0_domain) const
+    static double convert_active_time(uint64_t value, uint64_t &last_value, int &rollover_count)
```
```suggestion
    static double convert_active_time(uint64_t value, uint64_t &last_value, uint64_t &rollover_count)
```

bgeltz commented on libgeopmd/src/LevelZeroDevicePoolImp.hpp:
```diff
@@ -101,6 +102,8 @@ namespace geopm
             void check_domain_exists(int size, const char *func, int line) const;
             std::pair<unsigned int, unsigned int> subdevice_device_conversion(unsigned int idx) const;
+            mutable std::map<int, std::vector<uint64_t> > m_active_time_last; // Map from l0_domain to vector over gpu chips
+            mutable std::map<int, std::vector<int> > m_active_time_rollover; // Map from l0_domain to vector over gpu chips
```
```suggestion
            mutable std::map<int, std::vector<uint64_t> > m_active_time_rollover; // Map from l0_domain to vector over gpu chips
```

cmcantalupo committed:
Change rollover counter to uint64_t type

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
